### DISCUSSION TITLE
Is there any default retry middleware running?

### DISCUSSION THREAD

**Discussion by cemoktra** [tokio-rs/axum#1091] [Q&A]
I have an endpoint that does some heavy data gathering. When i call the endpoint once, i see the endpoint tracing three calls that are invoked exactly after one minute. And then curl times out after 3 minutes. My endpoint itself is not timing out, it never reaches the end. Also my metric middleware does not add any metric, meaning the inner call never succeeded or failed.

My conclusion is, this must be happening somewhere before that. I tried to look through the code but did not yet find anything. I'm not even sure if this relates to `axum`, `hyper` or even `tower`.

But something is going on that i would like to control from `axum`. Maybe somewhere a [budget](https://docs.rs/tower/latest/tower/retry/budget/struct.Budget.html) is added?

Any help or pointing to where some retry layer is added is appreciated.

Thanks

**Comment by davidpdrsn** [comment]
Nope axum doesn't add any such middleware by default.

Its hard to say without seeing the code. Are you able to make a minimal reproduction script that I can `cargo run`?

**Reply by cemoktra** [reply]
I can try next week, I'm off to long weekend without computer already

**Comment by cemoktra** [comment]
Ok the most simplest code (awaiting a sleep or a blocking sleep) does not reproduce this.

**Comment by TmLev** [comment]
I'm observing the same behaviour (3 retries) when trying to access a handler with `tower_http::timeout::TimeoutLayer` from a browser. `curl`, on the other hand, gave up after the first 408 error.

**Reply by davidpdrsn** [reply]
If you can make a script that reproduces it with minimal code then I can take a look.

**Reply by TmLev** [reply]
I'll try.

**Reply by TmLev** [reply]
Here you go: https://github.com/TmLev/axum-implicit-retry-repro/blob/master/src/main.rs

I tried three different browsers and now I'm actually starting to think that `axum` and `tower` has nothing to do with this issue. Looks like each browser has different retry policy (based on my attempts to reproduce aforementioned behaviour):

| Browser | Opening page for the first time | Consequent page refreshes |
| -- | -- | -- |
| Chrome | 1 attempt | 3 attempts |
| Yandex.Browser | 3 attempts | 3 attempts |
| Safari | 1 attempt | 1 attempt |

**Reply by davidpdrsn** [reply]
Yep this seems unrelated to axum and tower. I can confirm that I don't get retries in Safari.

**Reply by cemoktra** [reply]
With the exception that i did not have a timeout layer

**Reply by davidpdrsn** [reply]
Gimme a minimal example and I'll take a look.

**Reply by cemoktra** [reply]
I tried to generate one, but it was not reproducable. It might have to do with concurrency and blocking the handler
